src/BodyDigest.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+namespace HttpSignatures;
+
+class BodyDigest
+{
+    /** @var string */
+    private $hashName;
+
+    /** @var string */
+    private $digestHeaderPrefix;
+
+    /**
+     * @param string $name
+     * @return BodyDigest
+     * @throws DigestException
+     */
+    public function __construct($hashAlgorithm = null)
+    {
+        switch (strtolower(str_replace("-", '', $hashAlgorithm))) {
+        case 'sha':
+        case 'sha1':
+            $this->hashName = 'sha1';
+            $this->digestHeaderPrefix = 'SHA';
+            break;
+        case 'sha256':
+        case null:
+        case '':
+            $this->hashName = 'sha256';
+            $this->digestHeaderPrefix = 'SHA-256';
+            break;
+        case 'sha512':
+            $this->hashName = 'sha512';
+            $this->digestHeaderPrefix = 'SHA-512';
+            break;
+        default:
+            throw new DigestException("Digest algorithm parameter '$hashAlgorithm' not understood");
+            break;
+        }
+    }
+
+    public function putDigestInHeaderList($headerList)
+    {
+        if (!array_search('digest', $headerList->names)) {
+            $headerList->names[] = 'digest';
+        };
+        return $headerList;
+    }
+
+    public function setDigestHeader($message)
+    {
+        $message = $message->withoutHeader('Digest')
+            ->withHeader(
+                'Digest',
+                $this->getDigestHeaderLinefromBody($message->getBody()));
+        return $message;
+    }
+
+    public function getDigestHeaderLinefromBody($messageBody)
+    {
+        if (is_null($messageBody)) {
+            $messageBody = '';
+        };
```

Comment:
I'm not a fan of writing defensive code like this. If anything we should move to PHP 7+ and use a typed argument

Reply:
Until then, we still need a way to generate a digest that explicitly states "body is empty". Hashing a null is invalid, but hashing an empty string produces the right result. Until we go to PHP7, this is the best we can do I imagine?